src/chainrules.jl

```diff
@@ -0,0 +1,31 @@
+@non_differentiable validate_woodbury_arguments(A, D, S)
+
+function _times_pullback(Ȳ::AbstractMatrix, A, B, proj)
+    #Ā = @thunk(proj[:A](dot(Ȳ, B)'))
+    #B̄ = @thunk(proj[:B](A' * Ȳ))
+    Ā = dot(Ȳ, B)'
+    B̄ = A' * Ȳ
+    return (NoTangent(), Ā, B̄)
+end
+_times_pullback(ȳ::AbstractThunk, A, B, proj) = _times_pullback(unthunk(ȳ), A, B, proj)
+function _times_pullback(Ȳ::Tangent{<:WoodburyPDMat}, A, B, proj)
+    W = WoodburyPDMat(Ȳ.A, Ȳ.D, Ȳ.S)
```

Comment:
I still confused on why this is valid - does the tangent backing have to be able to reproduce the woodbury - i.e. considering the diagonals (which should be pos def) - their tangent is bound to positive? Again the PositiveReal float as an example.

Reply:
Not sure I follow - the pullback above would hit the `AbstractArray` method of `times_pullback`, right?

Reply:
Yes, but the point is on the tangent that is produced: 

```
julia> ∂W = unthunk(ChainRulesCore.rrule(*, D, W)[2](rand_tangent(D * W))[3])
Tangent{WoodburyPDMat{Float64, Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}}(A = [-0.26343438999522384 -1.3809461745889253; 2.8422893418996336 -0.13508471772276032; -10.328690745141067 -5.416957153117603; 5.441344472486843 3.0253900899189343], D = [-1.2857824438446035 0.0; 0.0 -4.303274675114356], S = [0.18734959448406613 0.0 0.0 0.0; 0.0 3.094752052943939 0.0 0.0; 0.0 0.0 -2.9860594258216753 0.0; 0.0 0.0 0.0 -0.26627135393506735])

julia> WoodburyPDMat(∂W.A, ∂W.D, ∂S)
ERROR: UndefVarError: ∂S not defined
Stacktrace:
 [1] top-level scope
   @ REPL[66]:1

julia> WoodburyPDMat(∂W.A, ∂W.D, ∂W.S)
ERROR: ArgumentError: Detected negative element on diagonal of D: [-1.2857824438446035 0.0; 0.0 -4.303274675114356]
```

Reply:
Oh, is that after the projection is applied? I didn't quite figure out whether that is correct

Reply:
There is afaict nothing wrong with the projection. This for example is passing (test_rrule(*, D, W)) and it is matching with the output of finite differences which doesn't use the projection. 

The point in question is again on the nature of the tangent. My understanding of your position is that it should always be valid to do `WoodburyPDMat(∂W.A, ∂W.D, ∂W.S)` and I don't understand this. My counterexample above is providing a tangent that would not fill this criteria. 

To try and fledge this out a bit, consider a type constrained to be positive real:

```
struct PositiveReal{T}
    x::T
    function PositiveReal(x::Real) = (@assert x > 0; new{T}(x))
end

and consider the pullback for this in any operation, say `*(A::Real, B::PositiveReal).` We have the pullbacks as `Ā = ΔΩ * B', B̄ = A' * ΔΩ`. If either A' or ΔΩ  are < 0  then B̄ < 0. This seems fine, as it's a gradient, but it does not mean that it can form a valid `PositiveReal`. Put another way the loss with respect to the element of B is negative, even though the primal is positive, so it doesn't make sense to try and do `PositiveReal{Tangent{PositiveReal}.x}`, which is effectively what `construct` in CR would try and do. To relate to my example above, I don't follow why it is necessary the case that `WoodburyPDMat(∂W.A, ∂W.D, ∂W.S)`  can be constructed (for example, with a negative element on the diagonals for the diagonals). 

Happy to accept if i've completely misunderstood something being less experienced in all this!

Reply:
Oh no you are absolutely right! I was just being really slow with understanding your point. Fleshing out the PositiveReal example cleared it up. I now understand your first point as well, sorry just didn't click the first time!